core/src/main/scala/kafka/coordinator/group/CoordinatorPartitionWriter.scala

```diff
@@ -201,18 +203,55 @@ class CoordinatorPartitionWriter[T](
     ))
   }
 
+  /**
+   * Verify the transaction.
+   *
+   * @param tp              The partition to write records to.
+   * @param transactionalId The transactional id.
+   * @param producerId      The producer id.
+   * @param producerEpoch   The producer epoch.
+   * @return A future containing the {@link VerificationGuard} or an exception.
+   * @throws KafkaException Any KafkaException caught during the operation.
+   */
+  override def maybeStartTransactionVerification(
+    tp: TopicPartition,
+    transactionalId: String,
+    producerId: Long,
+    producerEpoch: Short
+  ): CompletableFuture[VerificationGuard] = {
+    val future = new CompletableFuture[VerificationGuard]()
+    replicaManager.maybeStartTransactionVerificationForPartition(
+      topicPartition = tp,
+      transactionalId = transactionalId,
+      producerId = producerId,
+      producerEpoch = producerEpoch,
+      baseSequence = RecordBatch.NO_SEQUENCE,
+      requestLocal = RequestLocal.NoCaching,
+      callback = (error, _, verificationGuard) => {
+        if (error != Errors.NONE) {
+          future.completeExceptionally(error.exception)
+        } else {
+          future.complete(verificationGuard)
+        }
+      }
+    )
+    future
+  }
+
   private def internalAppend(
     tp: TopicPartition,
-    memoryRecords: MemoryRecords
+    memoryRecords: MemoryRecords,
+    verificationGuard: VerificationGuard = VerificationGuard.SENTINEL
   ): Long = {
     var appendResults: Map[TopicPartition, PartitionResponse] = Map.empty
-    replicaManager.appendRecords(
+    replicaManager.appendForGroup(
       timeout = 0L,
       requiredAcks = 1,
-      internalTopicsAllowed = true,
-      origin = AppendOrigin.COORDINATOR,
       entriesPerPartition = Map(tp -> memoryRecords),
       responseCallback = results => appendResults = results,
+      requestLocal = RequestLocal.NoCaching,
```

Comment:
do we not typically use request locals for this type of append?

Reply:
no, we don't have request locals here. the writer uses its own butter supplier (`threadLocalBufferSupplier`) so the `requestLocal` is not used at all.

Reply:
Right right, I remember this now. 
I'm wondering though is that ok with the async verification callback. Is it the case we don't evaluate which one to use until we execute the callback.

As an aside, did we decide that the callback is ok to execute on the request handler thread? I didn't go through and trace the threads being used yet.

Reply:
ah, right. i forgot about that one. let me check this tomorrow.

Reply:
Thanks! I will try to run through it today too.

Reply:
Yes -- that part has not changed. We "wrap" any callback from the transaction coordinator to the request handler thread.

Right now though, wrap only schedules to the request thread if we are already on a request thread. Otherwise we execute directly. If we start verification from a non-request handler thread, maybe this already works as you intend.

Alternatively, I could pass in a parameter to optionally wrap the callback (send it to the request thread) or not.

Reply:
> Right now though, wrap only schedules to the request thread if we are already on a request thread. Otherwise we execute directly. If we start verification from a non-request handler thread, maybe this already works as you intend.

My understanding is that `wrapAsyncCallback` will throw an exception if the caller is not a request thread and `bypassThreadCheck` is not set.

In my case, the verification is called from the request thread. I only re-schedule to the coordinator thread when the verification returns. I could also have scheduled the verification from the coordinator thread but it does not seem necessary.

> Alternatively, I could pass in a parameter to optionally wrap the callback (send it to the request thread) or not.

Yeah, that's an option. I played a bit with this idea this morning and the outcome was a bit confusing. Another option would be to do the re-scheduling on the caller side. I will continue to play... I think that we could address this in a subsequent PR to get rid of the extra hop. This PR works as-is but is suboptimal.

Reply:
Sorry -- you are correct. I was confused because we have a check for if we execute on the same request thread -- there we don't reschedule/wrap the request. (This was added when we error before we send the request to the txn coordinator)

Just for my understanding, right now we reschedule to the request thread, and that works fine -- the only concern is not using the allocated coordinator threads and taking up space on the request handler threads?

Just triple checking we won't run into an issue with request locals if the callback expects the buffer supplier from the coordinator thread.

Reply:
If it is the case its only an optimization question rather than a correctness one, I am good to tackle in a followup.

Reply:
Yes, we reschedule to request thread to only push an event to the coordinator thread. It is not optimal. The request local is not used at all as the writer uses its own buffers.

Reply:
Ok -- sorry I just meant I didn't know if it was passed through a callback (or future in this case) or if it was only invoked in the callback. For example, there were issues in the past since we used the buffer defined for the callback on a new handler thread.

Reply:
I guess we just complete the future in the callback and do the write in the thenCompose. That made me realize I don't know everything about how futures are scheduled, but looking at the docs I think it is safe to say this is correct.